dwave/optimization/symbols.pyx

```diff
@@ -2325,6 +2399,173 @@ cdef class NaryMultiply(ArraySymbol):
 _register(NaryMultiply, typeid(cppNaryMultiplyNode))
 
 
+cdef class NaryReduce(ArraySymbol):
+    """Using a supplied :class:`~dwave.optimization.model.Expression`, perform
+    a reduction operation along one or more array operands. The reduction
+    operation (represented by the ``Expression``) takes as input one value from
+    each of the operand arrays, as well as the result of the previously
+    computed operation, and computes a new value at the next output index.
+
+    This takes inspiration from
+    `numpy.ufunc.reduce <https://numpy.org/doc/2.1/reference/generated/numpy.ufunc.reduce.html>`_ 
+    but is different in that the reduction operation can take an arbitrary
+    number of arguments, instead of always two (hence the "n-ary" in
+    `NaryReduce`).
+
+    Args:
+        expression:
```

Comment:
I just noticed that [functools.reduce()](https://docs.python.org/3/library/functools.html#functools.reduce) and [std::accumulate()](https://en.cppreference.com/w/cpp/algorithm/accumulate) both use the left-most argument for the init/accumulated value. So I think we should follow that convention by making it the first input.